File: java/core/src/main/java/com/github/vigneshperiasami/persisto/client/Flowable.java
package com.github.vigneshperiasami.persisto.client;

public class Flowable<T> {
  private OnSubscribe<T> onSubscribe;

  public Flowable(OnSubscribe<T> onSubscribe) {
    this.onSubscribe = onSubscribe;
  }

  public final void listen(Subscriber<T> subscriber) {
    onSubscribe.call(new SafeSubscriber<>(subscriber));
  }

  public final <N> Flowable<N> lift(Operator<N, T> operator) {
    return new OperatorFlowable<>(this, operator);
  }

  public final Flowable<T> filter(final FuncB<T> filterFunc) {
    return new OperatorFlowable<>(this, new Operator<T, T>() {
      @Override
      public Subscriber<T> call(final Subscriber<T> nSubscriber) {
        return new Subscriber<T>() {
          @Override
          public void onNext(T data) {
            if (filterFunc.call(data)) {
              nSubscriber.onNext(data);
            }
          }

          @Override
          public void onError(Throwable err) {
            nSubscriber.onError(err);
          }
        };
      }
    });
  }

  public interface OnSubscribe<T> {
    void call(Subscriber<T> subscriber);
  }
}

File: node/core/flowable.js
"use strict";

class Flowable {
  constructor(listenFunc) {
    this.listenFunc = listenFunc;
  }

  lift(operatorFunc) {
    return new OperatorFlowable(this, operatorFunc);
  }

  filter(filterFunc) {
    return new OperatorFlowable(this, (onNext) => {
      return (data) => {
        if (filterFunc(data)) {
          onNext(data)
        }
      }
    })
  }

  listen(onNext) {
    this.listenFunc(onNext);
  }
}

class OperatorFlowable extends Flowable {
  constructor(flowable, operatorFunc) {
    super();
    this.flowable = flowable;
    this.operatorFunc = operatorFunc;
  }

  listen(onNext) {
    this.flowable.listen(this.operatorFunc(onNext));
  }
}

module.exports = (listenFunc) => {
  return new Flowable(listenFunc);
};
